File: URI/V1.2.8/javascript/src/conformance.mjs
/**
 * @aicent/rttp/conformance - the RTTP conformance vectors, and their replay.
 *
 * Why this package exists
 * -----------------------
 * A specification is worth exactly what an independent implementation can
 * reproduce from it. Two implementations that share no code, given the same
 * vectors, must produce bit-identical output; one implementation agreeing with
 * itself proves nothing.
 *
 * This module is therefore an **independent second implementation** of the same
 * vectors as `SPEC/tools/conformance.py` on the Python side. It deliberately
 * imports nothing from `src/rttp-uri.mjs` and must not: a conformance suite that
 * reuses the code it is checking only proves that the code equals itself.
 *
 * Authority boundary: URI syntax = RFC-002 sec. 10.2; frame layout = sec. 4.1 as
 * extended by `SPEC/RTTP-FRAME-EXT-v1.2.6.md`.
 *
 * Zero dependencies: `node:crypto` and `node:fs`, nothing else.
 */

import { createHash } from 'node:crypto';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

/**
 * The vector set shipped inside this package. Loading it with `readFileSync`
 * keeps this package importable from plain ESM without JSON import attributes,
 * and lets a caller replay a *newer* vector set without reinstalling.
 */
export const DEFAULT_VECTORS_PATH = fileURLToPath(new URL('./vectors.json', import.meta.url));

// ---------------------------------------------------------------------------
// Frame constants (sec. 4.1)
// ---------------------------------------------------------------------------

const MAGIC = 0x52545450;
const VERSION_ID = 130n;
const SIZE = 128;

/** Offset of ACTION_LEN and of the ACTION field itself. */
const OFF_ACTION_LEN = 0x68;
const OFF_ACTION = 0x69;

/** Offsets fixed by RTTP-FRAME-EXT-v1.2.6 sec. 2 (extension block at 0x66..0x7F). */
const OFF_SPEC_REV = 0x66;
const OFF_FLAGS = 0x67;
const OFF_RESERVED = 0x79;
const KNOWN_SPEC_REVS = new Set([0, 1]);
const FLAG_URI_ANCHORED = 0x01;

/** The ACTION field is 16 bytes; ACTION_LEN is one byte. See RFC-002 sec. 10.2. */
export const ACTION_MAX_LEN = 16;

const TOKEN = /^[a-z0-9-]+$/;
const HASH_INTENT = /^[0-9a-f]{8}$/;

// ---------------------------------------------------------------------------
// Independent implementation: URI parsing + canonicalisation
// ---------------------------------------------------------------------------

/**
 * Parse one `rttp` URI - an implementation written from RFC-002 sec. 10.2 alone,
 * sharing no code with `src/rttp-uri.mjs`.
 *
 * @param {string} uri
 * @returns {{ authority: string, action: string, shardHex: string }}
 */
export function parseUri(uri) {
  if (typeof uri !== 'string' || uri === '') throw new Error('empty uri');
  if (uri !== uri.trim()) throw new Error('leading/trailing whitespace');

  let rest;
  // The browser-registration form is accepted as a **tolerance only**: the
  // specification defines no such form, but a browser hands it to the handler.
  if (uri.startsWith('web+rttp://')) rest = uri.slice(11);
  else if (uri.startsWith('rttp://')) rest = uri.slice(7);
  else throw new Error('scheme must be rttp:// or the browser-registration form');

  for (const ch of ['@', '?', '#', '[', ']']) {
    if (uri.includes(ch)) throw new Error(`illegal char ${ch}`);
  }

  const slash = rest.indexOf('/');
  let authority;
  let action = '';
  if (slash === -1) {
    authority = rest;
  } else {
    authority = rest.slice(0, slash);
    action = rest.slice(slash + 1);
    if (action.includes('/')) throw new Error('path must be one segment');
    if (action === '') throw new Error('trailing slash with empty action');
  }

  const parts = authority.split('.');
  if (parts.length !== 3) throw new Error('authority must be intent.pillar.root');
  const [intent, pillar, root] = parts;

  const isHash = HASH_INTENT.test(intent);
  if (!isHash && !TOKEN.test(intent)) throw new Error('bad intent');
  if (!TOKEN.test(pillar)) throw new Error('bad pillar');
  if (!TOKEN.test(root)) throw new Error('bad root');
  if (action !== '' && !TOKEN.test(action)) throw new Error('bad action');

  for (const [value, name] of [[intent, 'intent'], [pillar, 'pillar'], [root, 'root']]) {
    if (value.startsWith('-') || value.endsWith('-')) {
      throw new Error(`${name} must not begin/end with '-'`);
    }
  }
  if (action !== '' && (action.startsWith('-') || action.endsWith('-'))) {
    throw new Error("action must not begin/end with '-'");
  }

  const canonicalAuthority = `${intent}.${pillar}.${root}`;
  const shard = createHash('sha256')
    .update(canonicalAuthority, 'ascii').digest().subarray(0, 16);

  return { authority: canonicalAuthority, action, shardHex: shard.toString('hex') };
}

// ---------------------------------------------------------------------------
// Independent implementation: the 128-byte frame
// ---------------------------------------------------------------------------

/**
 * Build one `PulseHeader128` from a URI and the vector's frame fields.
 *
 * Refuses an `action` that cannot fit the field rather than truncating it: a
 * silently shortened verb would be a routing decision made by the codec, and the
 * codec does not get to make that decision.
 *
 * @param {{
 *   sequenceId: number|string|bigint, ttl: number, priority: number,
 *   shardHex: string, aidHex: string, timestampNs: number|string|bigint,
 *   action: string, uriAnchored: boolean
 * }} fields
 * @returns {Buffer} 128 bytes
 */
export function buildFrame({ sequenceId, ttl, priority, shardHex, aidHex, timestampNs, action, uriAnchored }) {
  const actionBytes = Buffer.from(action, 'ascii');
  if (actionBytes.length > ACTION_MAX_LEN) {
    throw new Error(`action is ${actionBytes.length} bytes; the field holds ${ACTION_MAX_LEN}`);
  }
  if (shardHex.length !== 32) throw new Error('ROUTE_SHARD must be 32 hex characters');
  if (aidHex.length !== 64) throw new Error('AID_ORIGIN must be 64 hex characters');

  const buf = Buffer.alloc(SIZE);
  buf.writeUInt32BE(MAGIC, 0x00);
  buf.writeBigUInt64BE(0n, 0x04);                  // VERSION_ID u128 high half
  buf.writeBigUInt64BE(VERSION_ID, 0x0C);          // VERSION_ID = 130
  const seq = BigInt(sequenceId);
  buf.writeBigUInt64BE(seq >> 64n, 0x14);          // SEQUENCE_ID u128
  buf.writeBigUInt64BE(seq & 0xFFFFFFFFFFFFFFFFn, 0x1C);
  const ts = BigInt(timestampNs);
  buf.writeBigUInt64BE(ts >> 64n, 0x24);           // TIMESTAMP u128
  buf.writeBigUInt64BE(ts & 0xFFFFFFFFFFFFFFFFn, 0x2C);
  buf.writeUInt8(ttl & 0xFF, 0x34);                // TTL_PULSE
  buf.writeUInt8(priority & 0xFF, 0x35);           // PRIORITY
  Buffer.from(shardHex, 'hex').copy(buf, 0x36);    // ROUTE_SHARD u128
  Buffer.from(aidHex, 'hex').copy(buf, 0x46);      // AID_ORIGIN 256-bit
  buf[0x66] = 1;                                   // SPEC_REV
  buf[0x67] = uriAnchored ? 1 : 0;                 // FLAGS
  buf[OFF_ACTION_LEN] = actionBytes.length;        // ACTION_LEN
  actionBytes.copy(buf, OFF_ACTION);               // ACTION (zero-padded)
  return buf;                                      // 0x79..0x80 RESERVED = 0
}

/**
 * Verify a `PulseHeader128`, mirroring the Python `pulse_header.verify`.
 *
 * Implements the extension-block rules of RTTP-FRAME-EXT-v1.2.6 sec. 2 / sec. 4
 * (R2-R7) and the ACTION grammar of RFC-002 sec. 10.2. Every failure throws: a
 * check that cannot be performed is a failure, never a pass.
 *
 * @param {Buffer} raw 128 bytes
 * @param {{ expectedAidOrigin?: string }} [options]
 * @returns {{ specRev: number, flags: number, uriAnchored: boolean,
 *             action: string, actionOmitted: boolean }}
 */
export function verifyFrame(raw, options = {}) {
  if (!Buffer.isBuffer(raw) || raw.length !== SIZE) {
    throw new Error(`frame must be ${SIZE} bytes`);
  }
  if (raw.readUInt32BE(0x00) !== MAGIC) throw new Error('bad RTTP_MAGIC');
  if (raw.readBigUInt64BE(0x0C) !== VERSION_ID) throw new Error('unsupported VERSION_ID');

  const specRev = raw[OFF_SPEC_REV];
  if (!KNOWN_SPEC_REVS.has(specRev)) {
    throw new Error(`unknown SPEC_REV: ${specRev} (fail closed)`);
  }

  const flags = raw[OFF_FLAGS];
  const actionLen = raw[OFF_ACTION_LEN];

  if (specRev === 0) {
    // R6: a pre-v1.2.6 frame must have an all-zero extension block.
    for (let i = OFF_SPEC_REV; i < SIZE; i += 1) {
      if (raw[i] !== 0) throw new Error('SPEC_REV=0 but extension block is not all-zero');
    }
  } else {
    // R7: unknown FLAGS bits and any RESERVED byte are rejections.
    if ((flags & ~FLAG_URI_ANCHORED) !== 0) {
      throw new Error(`unknown FLAGS bits set: 0x${(flags & ~FLAG_URI_ANCHORED).toString(16).padStart(2, '0')}`);
    }
    for (let i = OFF_RESERVED; i < SIZE; i += 1) {
      if (raw[i] !== 0) throw new Error('RESERVED bytes must be zero in SPEC_REV=1');
    }
    // R4: ACTION_LEN is one byte, the field holds 16.
    if (actionLen > ACTION_MAX_LEN) {
      throw new Error(`ACTION_LEN ${actionLen} > ${ACTION_MAX_LEN}`);
    }
    for (let i = OFF_ACTION + actionLen; i < OFF_ACTION + ACTION_MAX_LEN; i += 1) {
      if (raw[i] !== 0) throw new Error('ACTION padding must be zero');
    }
  }

  let action = '';
  if (specRev === 1 && actionLen > 0) {
    action = raw.subarray(OFF_ACTION, OFF_ACTION + actionLen).toString('ascii');
    if (!TOKEN.test(action)) {
      throw new Error(`ACTION violates RFC-002 sec. 10.2: ${JSON.stringify(action)}`);
    }
  }

  if (options.expectedAidOrigin !== undefined) {
    const got = raw.subarray(0x46, 0x66).toString('hex');
    if (got !== options.expectedAidOrigin) {
      throw new Error('AID_ORIGIN does not match the expected origin');
    }
  }

  return {
    specRev,
    flags,
    uriAnchored: (flags & FLAG_URI_ANCHORED) !== 0,
    action,
    actionOmitted: action === '',
  };
}

// ---------------------------------------------------------------------------
// Vector loading and replay
// ---------------------------------------------------------------------------

/**
 * @param {string} [path] defaults to the vector set shipped with this package
 * @returns {object} the parsed vector set
 */
export function loadVectors(path = DEFAULT_VECTORS_PATH) {
  return JSON.parse(readFileSync(path, 'utf8'));
}

/**
 * Replay a vector set against the independent implementations above.
 *
 * @param {object} vectors
 * @returns {{ checked: number, failures: string[], passed: boolean }}
 */
export function runConformance(vectors) {
  const failures = [];
  let checked = 0;

  for (const c of vectors.positive_uris) {
    checked += 1;
    try {
      const got = parseUri(c.uri);
      if (got.authority !== c.authority) failures.push(`${c.uri} :: authority ${got.authority} != ${c.authority}`);
      if (got.action !== c.action) failures.push(`${c.uri} :: action ${got.action} != ${c.action}`);
      if (got.shardHex !== c.route_shard_hex) failures.push(`${c.uri} :: shard ${got.shardHex} != ${c.route_shard_hex}`);
    } catch (err) {
      failures.push(`${c.uri} :: expected ACCEPT but threw: ${err.message}`);
    }
  }

  for (const c of vectors.negative_uris) {
    checked += 1;
    try {
      parseUri(c.uri);
      failures.push(`${c.uri} :: expected REJECT but parsed`);
    } catch { /* expected */ }
  }

  const aidHex = vectors.frame_vectors[0].aid_origin_hex;

  for (const c of vectors.frame_vectors) {
    checked += 1;
    const { shardHex, action } = parseUri(c.uri);
    const frame = buildFrame({
      sequenceId: c.sequence_id,
      ttl: c.ttl,
      priority: c.priority,
      shardHex,
      aidHex,
      timestampNs: c.timestamp_ns,
      action,
      uriAnchored: true,
    });
    if (frame.toString('hex') !== c.header_hex) {
      failures.push(`${c.uri} :: frame bytes mismatch`);
    }
    try {
      verifyFrame(frame, { expectedAidOrigin: aidHex });
    } catch (err) {
      failures.push(`${c.uri} :: published frame rejected by verifyFrame: ${err.message}`);
    }
  }

  for (const c of (vectors.negative_frames || [])) {
    checked += 1;
    try {
      verifyFrame(Buffer.from(c.header_hex, 'hex'), { expectedAidOrigin: aidHex });
      failures.push(`frame ${c.rule} :: ${c.label} :: expected REJECT but accepted`);
    } catch { /* expected */ }
  }

  for (const c of (vectors.positive_frames || [])) {
    checked += 1;
    try {
      const got = verifyFrame(Buffer.from(c.header_hex, 'hex'), { expectedAidOrigin: aidHex });
      if (got.action !== c.action) {
        failures.push(`frame ${c.rule} :: action ${got.action} != ${c.action}`);
      }
    } catch (err) {
      failures.push(`frame ${c.rule} :: ${c.label} :: expected ACCEPT but rejected: ${err.message}`);
    }
  }

  {
    checked += 1;
    const cv = vectors.compat_vector_spec_rev_0;
    const { shardHex, action } = parseUri(cv.uri);
    const frame = buildFrame({
      sequenceId: vectors.frame_vectors[0].sequence_id,
      ttl: vectors.frame_vectors[0].ttl,
      priority: vectors.frame_vectors[0].priority,
      shardHex,
      aidHex,
      timestampNs: vectors.frame_vectors[0].timestamp_ns,
      action,
      uriAnchored: true,
    });
    frame.fill(0, 0x66, SIZE);  // SPEC_REV=0: the extension block is all zero
    if (frame.toString('hex') !== cv.header_hex) {
      failures.push('compat vector :: frame bytes mismatch');
    }
    try {
      const got = verifyFrame(Buffer.from(cv.header_hex, 'hex'), { expectedAidOrigin: aidHex });
      if (got.specRev !== 0 || !got.actionOmitted) {
        failures.push('compat vector :: SPEC_REV=0 semantics mismatch');
      }
    } catch (err) {
      failures.push(`compat vector :: verifyFrame rejected an old frame: ${err.message}`);
    }
  }

  return { checked, failures, passed: failures.length === 0 };
}

File: URI/V1.2.8/rust/src/conformance.rs
//! Conformance replay -- the published vector set, executed against THIS build.
//!
//! The vectors ship inside the crate (`vectors/rttp-conformance-v1.2.6.json`)
//! so a stranger can verify an installed package offline, with no repository
//! checkout. The replay covers every published check:
//!   6 positive URIs + 15 negative URIs + 3 frame vectors + 1 compat vector
//!   + 9 fail-closed frames + 1 accepted frame (SPEC_REV extension, sec. 3.2/4)
//!   + 3 envelope checks (canonical signing input, AID self-certification,
//!     Ed25519 signature).
//!
//! The JSON reader (`crate::json`) and the grammar (`crate::rttp_uri`) share
//! no code paths beyond the public parse function -- the replay never trusts
//! a helper written "by the same hand" for a different layer. The envelope's
//! canonical signing input is rebuilt here, from the vector, on purpose: the
//! seal module's own helper is never used to check the seal module.
//!
//! **Nothing may be skipped silently.** A zero-dependency build cannot do the
//! Ed25519 arithmetic, so that one check is reported as `skipped` and counted
//! -- the tests below assert `checked + skipped == PUBLISHED_CHECKS`, so a
//! check can not disappear from the report by accident.

use crate::json::{parse as json_parse, Json};
use crate::pulse_header;
use crate::rttp_uri;
use crate::sha256::{bytes_to_hex, hex_to_bytes, sha256};

/// Domain separation prefix of the seal envelope (SPEC/RTTP-SEAL-ENVELOPE
/// v1.2.6 sec. 3). Deliberately repeated here instead of imported from
/// `crate::seal_asym`: the replay must read the wire format, not the module
/// that writes it.
const SEAL_DOMAIN: &[u8] = b"rttp-seal-v1\n";

/// Checks the published vector set contains in total.
pub const PUBLISHED_CHECKS: usize = 38;

/// Checks a build without the `ed25519` feature can execute -- 38 minus the
/// signature arithmetic.
pub const CHECKS_WITHOUT_PRIMITIVE: usize = 37;

/// Result of one full replay.
#[derive(Debug, Clone)]
pub struct ConformanceResult {
    pub checked: usize,
    pub passed: usize,
    /// Checks this build could not run (optional primitive absent). Always
    /// visible in the count -- never a quiet zero.
    pub skipped: usize,
    pub failures: Vec<String>,
}

impl ConformanceResult {
    /// One-line summary, same shape the Python and JavaScript packages print.
    pub fn summary(&self) -> String {
        if !self.failures.is_empty() {
            format!(
                "[FAIL] {} of {} checks failed ({} skipped)",
                self.failures.len(),
                self.checked,
                self.skipped
            )
        } else if self.skipped > 0 {
            format!(
                "[PASS] all {} checks passed ({} skipped)",
                self.checked, self.skipped
            )
        } else {
            format!("[PASS] all {} checks passed", self.checked)
        }
    }
}

/// Rebuild the canonical signing input and the self-certifying AID from the
/// envelope vector, using only the JSON reader and the in-crate SHA-256.
fn signing_input_from_vector(ev: &Json) -> Result<(Vec<u8>, String), String> {
    let payload = ev.get("payload").cloned().ok_or("missing payload")?;
    if !matches!(payload, Json::Obj(_)) {
        return Err("payload must be an object".into());
    }
    let field = |key: &str| {
        ev.get(key)
            .cloned()
            .ok_or_else(|| format!("missing {}", key))
    };
    let inner = Json::Obj(vec![
        ("aid".into(), field("aid")?),
        ("alg".into(), field("alg")?),
        ("nonce".into(), field("nonce")?),
        ("payload".into(), payload),
        ("pub".into(), field("pub")?),
        ("ts".into(), field("ts")?),
    ]);
    let mut input = SEAL_DOMAIN.to_vec();
    input.extend_from_slice(inner.canonical().as_bytes());
    let pub_hex = ev.get("pub").and_then(Json::as_str).ok_or("missing pub")?;
    let pub_bytes = hex_to_bytes(pub_hex).ok_or("pub is not hex")?;
    if pub_bytes.len() != 32 {
        return Err("pub must be 32 bytes".into());
    }
    Ok((input, bytes_to_hex(&sha256(&pub_bytes))))
}

/// Rebuild the complete envelope object from the vector, for the signature
/// check. Only needed with the `ed25519` feature.
#[cfg(feature = "ed25519")]
fn envelope_object_from_vector(ev: &Json) -> Option<Json> {
    let text = |key: &str| ev.get(key).and_then(Json::as_str).map(str::to_string);
    Some(Json::Obj(vec![
        ("v".into(), Json::Num("1".into())),
        ("alg".into(), Json::Str(text("alg")?)),
        ("aid".into(), Json::Str(text("aid")?)),
        ("pub".into(), Json::Str(text("pub")?)),
        ("ts".into(), ev.get("ts")?.clone()),
        ("nonce".into(), Json::Str(text("nonce")?)),
        ("payload".into(), ev.get("payload")?.clone()),
        ("sig".into(), Json::Str(text("sig")?)),
    ]))
}

/// Replay a parsed vector document.
pub fn run(doc: &Json) -> ConformanceResult {
    let mut res = ConformanceResult {
        checked: 0,
        passed: 0,
        skipped: 0,
        failures: Vec::new(),
    };

    let check = |res: &mut ConformanceResult, ok: bool, label: &str, detail: String| {
        res.checked += 1;
        if ok {
            res.passed += 1;
        } else {
            res.failures.push(format!("{}: {}", label, detail));
        }
    };

    // --- positive URIs: accept, canonicalise, derive ------------------------
    if let Some(cases) = doc.get("positive_uris").and_then(Json::as_arr) {
        for case in cases {
            let uri = case.get("uri").and_then(Json::as_str).unwrap_or("");
            let parsed = rttp_uri::parse(uri);
            match parsed {
                Err(e) => check(
                    &mut res,
                    false,
                    uri,
                    format!("expected ACCEPT, rejected: {}", e),
                ),
                Ok(p) => {
                    let mut bad: Vec<String> = Vec::new();
                    for (key, got, want) in [
                        (
                            "canonical_uri",
                            p.canonical_uri.clone(),
                            case.get("canonical_uri")
                                .and_then(Json::as_str)
                                .unwrap_or("")
                                .to_string(),
                        ),
                        (
                            "action",
                            p.action.clone(),
                            case.get("action")
                                .and_then(Json::as_str)
                                .unwrap_or("")
                                .to_string(),
                        ),
                        (
                            "route_shard_hex",
                            crate::sha256::bytes_to_hex(&p.route_shard),
                            case.get("route_shard_hex")
                                .and_then(Json::as_str)
                                .unwrap_or("")
                                .to_string(),
                        ),
                    ] {
                        if got != want {
                            bad.push(format!("{} {:?} != {:?}", key, want, got));
                        }
                    }
                    check(&mut res, bad.is_empty(), uri, bad.join("; "));
                }
            }
        }
    }

    // --- negative URIs: reject, fail closed ---------------------------------
    if let Some(cases) = doc.get("negative_uris").and_then(Json::as_arr) {
        for case in cases {
            let uri = case.get("uri").and_then(Json::as_str).unwrap_or("");
            check(
                &mut res,
                rttp_uri::parse(uri).is_err(),
                uri,
                "expected REJECT but parsed".into(),
            );
        }
    }

    // --- frame vectors: URI -> ROUTE_SHARD -> 128 bytes, bit for bit ----------
    if let Some(cases) = doc.get("frame_vectors").and_then(Json::as_arr) {
        for case in cases {
            let uri = case.get("uri").and_then(Json::as_str).unwrap_or("");
            let label = format!("frame {}", uri);
            let fail = |res: &mut ConformanceResult, detail: String| {
                check(res, false, &label, detail);
            };
            let Some(aid_hex) = case.get("aid_origin_hex").and_then(Json::as_str) else {
                fail(&mut res, "missing aid_origin_hex".into());
                continue;
            };
            let Some(aid) = hex_to_bytes(aid_hex) else {
                fail(&mut res, "bad aid_origin_hex".into());
                continue;
            };
            let mut aid32 = [0u8; 32];
            if aid.len() != 32 {
                fail(&mut res, "aid_origin_hex must be 32 bytes".into());
                continue;
            }
            aid32.copy_from_slice(&aid);

            let seq = case.get("sequence_id").and_then(Json::as_u128).unwrap_or(0);
            let ttl = case.get("ttl").and_then(Json::as_u128).unwrap_or(0) as u8;
            let prio = case.get("priority").and_then(Json::as_u128).unwrap_or(0) as u8;
            let ts = case
                .get("timestamp_ns")
                .and_then(Json::as_u128)
                .unwrap_or(0);
            let want_hex = case.get("header_hex").and_then(Json::as_str).unwrap_or("");

            match pulse_header::build_for_uri(seq, ttl, prio, uri, &aid32, ts) {
                Err(e) => fail(&mut res, format!("build failed: {}", e)),
                Ok(raw) => {
                    if crate::sha256::bytes_to_hex(&raw) != want_hex {
                        fail(
                            &mut res,
                            "frame bytes differ from the published vector".into(),
                        );
                        continue;
                    }
                    match pulse_header::verify(&raw, Some(&aid32)) {
                        Err(e) => fail(&mut res, format!("verify failed on our own frame: {}", e)),
                        Ok(fields) => {
                            let want_action =
                                case.get("action").and_then(Json::as_str).unwrap_or("");
                            check(
                                &mut res,
                                fields.action == want_action,
                                &label,
                                format!("action readback {:?} != {:?}", fields.action, want_action),
                            );
                        }
                    }
                }
            }
        }
    }

    // --- fail-closed frames: malformed extension blocks must be rejected -----
    // Each vector names the rule it exercises (R4-R7 of SPEC sec. 4); the
    // origin is embedded in the frame itself, so nothing is pinned here.
    if let Some(cases) = doc.get("negative_frames").and_then(Json::as_arr) {
        for case in cases {
            let label = case
                .get("label")
                .and_then(Json::as_str)
                .unwrap_or("negative frame");
            let raw = case
                .get("header_hex")
                .and_then(Json::as_str)
                .and_then(hex_to_bytes);
            let Some(bytes) = raw else {
                check(&mut res, false, label, "missing or bad header_hex".into());
                continue;
            };
            check(
                &mut res,
                pulse_header::verify(&bytes, None).is_err(),
                label,
                "expected REJECT but the frame verified".into(),
            );
        }
    }

    // --- accepted frame: SPEC_REV=1 with ACTION_LEN=0 == omission ------------
    if let Some(cases) = doc.get("positive_frames").and_then(Json::as_arr) {
        for case in cases {
            let label = case
                .get("label")
                .and_then(Json::as_str)
                .unwrap_or("accepted frame");
            let raw = case
                .get("header_hex")
                .and_then(Json::as_str)
                .and_then(hex_to_bytes);
            let Some(bytes) = raw else {
                check(&mut res, false, label, "missing or bad header_hex".into());
                continue;
            };
            let want_action = case.get("action").and_then(Json::as_str).unwrap_or("");
            match pulse_header::verify(&bytes, None) {
                Err(e) => check(
                    &mut res,
                    false,
                    label,
                    format!("expected ACCEPT, rejected: {}", e),
                ),
                Ok(fields) => check(
                    &mut res,
                    fields.action == want_action,
                    label,
                    format!("action readback {:?} != {:?}", fields.action, want_action),
                ),
            }
        }
    }

    // --- envelope vector ----------------------------------------------------
    // Two checks run in every build; the signature is the third and the only
    // thing a zero-dependency build may skip.
    if let Some(ev) = doc.get("envelope_vector") {
        let want_input = ev
            .get("signing_input_hex")
            .and_then(Json::as_str)
            .unwrap_or("");
        let want_aid = ev.get("aid").and_then(Json::as_str).unwrap_or("");

        match signing_input_from_vector(ev) {
            Err(e) => {
                check(
                    &mut res,
                    false,
                    "envelope: canonical signing input",
                    e.clone(),
                );
                check(&mut res, false, "envelope: AID self-certification", e);
            }
            Ok((input, aid)) => {
                let got_input = bytes_to_hex(&input);
                check(
                    &mut res,
                    got_input == want_input,
                    "envelope: canonical signing input",
                    format!("{:?} != {:?}", want_input, got_input),
                );
                check(
                    &mut res,
                    aid == want_aid,
                    "envelope: AID self-certification",
                    format!("{:?} != {:?}", want_aid, aid),
                );
            }
        }

        #[cfg(feature = "ed25519")]
        {
            // The vector is pinned to a fixed ts, so the freshness window is
            // deliberately not enforced here.
            match envelope_object_from_vector(ev) {
                None => check(
                    &mut res,
                    false,
                    "envelope: Ed25519 signature",
                    "envelope vector is incomplete".into(),
                ),
                Some(env) => match crate::seal_asym::verify_envelope(&env, None, false) {
                    Ok(aid) => check(
                        &mut res,
                        aid == want_aid,
                        "envelope: Ed25519 signature",
                        format!("verified AID {:?} != {:?}", aid, want_aid),
                    ),
                    Err(e) => check(
                        &mut res,
                        false,
                        "envelope: Ed25519 signature",
                        format!("expected ACCEPT, rejected: {}", e),
                    ),
                },
            }
        }
        #[cfg(not(feature = "ed25519"))]
        {
            res.skipped += 1;
        }
    }

    // --- compat vector: SPEC_REV=0 frames must still verify ------------------
    if let Some(cv) = doc.get("compat_vector_spec_rev_0") {
        let label = "compat SPEC_REV=0";
        let ok = (|| -> Result<bool, String> {
            let hex = cv
                .get("header_hex")
                .and_then(Json::as_str)
                .ok_or("missing header_hex")?;
            let raw = hex_to_bytes(hex).ok_or("bad header_hex")?;
            let aid_hex = cv.get("uri").and_then(Json::as_str).unwrap_or("");
            let _ = aid_hex;
            // The compat vector's aid_origin is inside the frame itself --
            // verify without pinning, then require the rev-0 semantics.
            let fields = pulse_header::verify(&raw, None).map_err(|e| e.to_string())?;
            Ok(fields.spec_rev == 0 && fields.action_omitted)
        })();
        match ok {
            Ok(true) => check(&mut res, true, label, String::new()),
            Ok(false) => check(
                &mut res,
                false,
                label,
                "SPEC_REV=0 semantics mismatch".into(),
            ),
            Err(e) => check(
                &mut res,
                false,
                label,
                format!("expected ACCEPT, rejected: {}", e),
            ),
        }
    }

    res
}

/// Convenience: parse the shipped vector file and replay it.
pub fn run_shipped() -> ConformanceResult {
    match json_parse(include_str!("../vectors/rttp-conformance-v1.2.6.json")) {
        Ok(doc) => run(&doc),
        Err(e) => ConformanceResult {
            checked: 1,
            passed: 0,
            skipped: 0,
            failures: vec![format!("cannot parse shipped vectors: {}", e)],
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shipped_vectors_replay_green() {
        let r = run_shipped();
        println!("{}", r.summary());
        assert!(r.failures.is_empty(), "failures: {:?}", r.failures);
        assert_eq!(
            r.checked + r.skipped,
            PUBLISHED_CHECKS,
            "the whole published vector set must be replayed -- nothing may vanish silently"
        );
        assert_eq!(
            r.checked,
            CHECKS_WITHOUT_PRIMITIVE + usize::from(cfg!(feature = "ed25519")),
            "only the Ed25519 primitive itself may be skipped in the default build"
        );
        assert_eq!(r.checked, r.passed);
    }
}
